### PR TITLE
Feature/core/subscribe error handling

### PR DESCRIPTION
**Pull request by TalhaAli00**
Added some internal error handling based on finding from Elrond Wallet

### CODE REVIEW

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt [hunk 1]
```diff
@@ -46,7 +45,11 @@ internal object PairingClient : PairingInterface {
                     launch(Dispatchers.IO) {
                         pairingRepository.getListOfPairings()
                             .map { pairing -> pairing.topic }
-                            .onEach { pairingTopic -> jsonRpcInteractor.subscribe(pairingTopic) }
+                            .onEach { pairingTopic ->
+                                try {
+                                    jsonRpcInteractor.subscribe(pairingTopic)
+                                } catch (_: NoRelayConnectionException) {}
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt:51]: What's the flow when we check this exception here? Not action, why?

File: auth/sdk/src/main/kotlin/com/walletconnect/auth/engine/domain/AuthEngine.kt [hunk 2]
```diff
@@ -226,7 +233,9 @@ internal class AuthEngine(
         pairingTopicToResponseTopicMap
             .map { it.value }
             .onEach { responseTopic: Topic ->
-                jsonRpcInteractor.subscribe(responseTopic)
+                try {
+                    jsonRpcInteractor.subscribe(responseTopic)
+                } catch (_: NoRelayConnectionException) {}
```
Comment (jakobuid) [inline auth/sdk/src/main/kotlin/com/walletconnect/auth/engine/domain/AuthEngine.kt:238]: Same

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt [hunk 3]
```diff
@@ -724,6 +730,7 @@ onFailure(error)
                 val approveParams = response.result as SignParamsVO.ApprovalParams
                 val responderPublicKey = PublicKey(approveParams.responderPublicKey)
                 val sessionTopic = crypto.generateTopicFromKeyAgreement(selfPublicKey, responderPublicKey)
+                // TODO: Should we emit an internal error here?
```
Comment (jakobuid) [inline sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt:733]: Imo, why not?

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt [hunk 4]
```diff
@@ -820,7 +827,12 @@ onFailure(error)
             }
 
         listOfValidSessions
-            .onEach { session -> jsonRpcInteractor.subscribe(session.topic) }
+            .onEach { session ->
+                try {
+                    jsonRpcInteractor.subscribe(session.topic)
+                } catch (_: NoRelayConnectionException) {
+                }
```
Comment (jakobuid) [inline sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt:834]: Same

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt [hunk 5]
```diff
@@ -46,7 +45,11 @@ internal object PairingClient : PairingInterface {
                     launch(Dispatchers.IO) {
                         pairingRepository.getListOfPairings()
                             .map { pairing -> pairing.topic }
-                            .onEach { pairingTopic -> jsonRpcInteractor.subscribe(pairingTopic) }
+                            .onEach { pairingTopic ->
+                                try {
+                                    jsonRpcInteractor.subscribe(pairingTopic)
+                                } catch (_: Exception) {}
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt:51]: I think we should at least log an error or emit internal error to client's delegate

File: chat/sdk/src/main/kotlin/com/walletconnect/chat/engine/domain/ChatEngine.kt [hunk 6]
```diff
@@ -35,21 +39,44 @@ internal class ChatEngine(
...
     private val _events: MutableSharedFlow<EngineEvent> = MutableSharedFlow()
     val events: SharedFlow<EngineEvent> = _events.asSharedFlow()
     private val inviteRequestMap: MutableMap<Long, WCRequest> = mutableMapOf()
 
     init {
-        collectJsonRpcRequests()
-        collectPeerResponses()
-        resubscribeToInviteTopic()
-        collectInternalErrors()
+        pairingInterface.register(
```
Comment (jakobuid) [inline chat/sdk/src/main/kotlin/com/walletconnect/chat/engine/domain/ChatEngine.kt:50]: Why does chat sdk need register?

File: chat/sdk/src/main/kotlin/com/walletconnect/chat/engine/domain/ChatEngine.kt [hunk 7]
```diff
@@ -344,27 +371,44 @@ internal class ChatEngine(
...
+            }
             Logger.log("Listening for invite on: $topic, pubKey X:$publicKey")
         } catch (error: Exception) {
             scope.launch { _events.emit(SDKError(InternalError(error))) }
         }
     }
 
     private fun collectInternalErrors() {
-        jsonRpcInteractor.internalErrors
+        merge(jsonRpcInteractor.internalErrors, pairingInterface.findWrongMethodsFlow)
```
Comment (jakobuid) [inline chat/sdk/src/main/kotlin/com/walletconnect/chat/engine/domain/ChatEngine.kt:411]: Why does chat sdk need `findWrongMethodsFlow`?

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/common/model/vo/clientsync/session/SignRpcVO.kt [hunk 8]
```diff
@@ -5,14 +5,14 @@ package com.walletconnect.sign.common.model.vo.clientsync.session
 import com.squareup.moshi.Json
 import com.squareup.moshi.JsonClass
 import com.walletconnect.android.internal.common.model.JsonRpcClientSync
-import com.walletconnect.sign.common.model.vo.clientsync.session.params.SignParamsVO
+import com.walletconnect.sign.common.model.vo.clientsync.session.params.SignParams
 import com.walletconnect.sign.json_rpc.model.JsonRpcMethod
 
-internal sealed class SignRpcVO : JsonRpcClientSync<SignParamsVO> {
+internal sealed class SignRpcVO : JsonRpcClientSync<SignParams> {
```
Comment (jakobuid) [inline sign/sdk/src/main/kotlin/com/walletconnect/sign/common/model/vo/clientsync/session/SignRpcVO.kt:11]: SignRpcVO -> SignRpc ?

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt [hunk 9]
```diff
@@ -722,10 +732,20 @@ internal class SignEngine(
             is JsonRpcResponse.JsonRpcResult -> {
                 Logger.log("Session proposal approve received")
                 val selfPublicKey = PublicKey(params.proposer.publicKey)
-                val approveParams = response.result as SignParamsVO.ApprovalParams
+                val approveParams = response.result as SignParams.ApprovalParams
                 val responderPublicKey = PublicKey(approveParams.responderPublicKey)
                 val sessionTopic = crypto.generateTopicFromKeyAgreement(selfPublicKey, responderPublicKey)
-                jsonRpcInteractor.subscribe(sessionTopic)
+                try {
```
Comment (jakobuid) [inline sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt:738]: Do we need a try catch when calling subscribe method? From code it looks like all errors from subscribe are returned using callback

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/common/model/vo/clientsync/session/SignRpcVO.kt (hunk 8 above)
Comment (TalhaAli00) [reply sign/sdk/src/main/kotlin/com/walletconnect/sign/common/model/vo/clientsync/session/SignRpcVO.kt:11]: renamed

File: androidCore/impl/src/main/kotlin/com/walletconnect/android/impl/json_rpc/domain/JsonRpcInteractor.kt [hunk 10]
```diff
@@ -158,7 +168,7 @@ internal class JsonRpcInteractor(
     }
 
     override fun subscribe(topic: Topic) {
-        checkConnectionWorking()
+        checkConnectionWorking() // TODO: Should we add an onFailure callback?
```
[the remark this answers is not in the document]
Comment (TalhaAli00) [reply androidCore/impl/src/main/kotlin/com/walletconnect/android/impl/json_rpc/domain/JsonRpcInteractor.kt]: addressed

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt (hunk 1 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt]: added a new flow

File: auth/sdk/src/main/kotlin/com/walletconnect/auth/engine/domain/AuthEngine.kt (hunk 2 above)
Comment (TalhaAli00) [reply auth/sdk/src/main/kotlin/com/walletconnect/auth/engine/domain/AuthEngine.kt]: addressed

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt (hunk 3 above)
Comment (TalhaAli00) [reply sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt]: removed comment

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt (hunk 4 above)
Comment (TalhaAli00) [reply sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt:855]: addressed

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt (hunk 5 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/pairing/PairingClient.kt:51]: added internal error handling

**Review by jakobuid — APPROVED** [review]

**Pull request merged**